dht-cache/src/cache.rs

```diff
@@ -0,0 +1,462 @@
+        self.local.put(&elem).await;
+
+        self.cmd
+            .send(Command::Publish(serde_json::to_value(&elem)?))
+            .map_err(|_| Error::Channel)?;
+
+        Ok(())
+    }
+
+    /// Query the local cache
+    pub fn query(&self, topic: &str) -> Query {
+        self.local.query(topic)
+    }
+
+    /// Get a list of the current peers
+    pub async fn peers(&self) -> Vec<PeerInfo> {
+        let peers = self.peers.read().await;
+        peers
+            .list
+            .values()
+            .map(|p| PeerInfo {
+                peer_id: p.peer_id.to_owned(),
+                hash: p.cache_hash,
+                last_seen: p.publication_timestamp,
+            })
+            .collect()
+    }
+
+    /// Return the current cache hash
+    pub async fn get_hash(&self) -> u64 {
+        self.local.get_hash().await
+    }
+}
+
+#[derive(Default, Debug, Clone)]
+pub(crate) struct PeersState {
+    list: BTreeMap<String, DomoCacheStateMessage>,
+    last_repub_timestamp: u128,
+    repub_interval: u128,
+}
+
+#[derive(Debug)]
+enum CacheState {
+    Synced,
+    Desynced { is_leader: bool },
+}
+
+impl PeersState {
+    fn with_interval(repub_interval: u128) -> Self {
+        Self {
+            repub_interval,
+            ..Default::default()
+        }
+    }
+
+    fn insert(&mut self, state: DomoCacheStateMessage) {
+        self.list.insert(state.peer_id.to_string(), state);
+    }
+
+    fn is_synchronized(&self, peer_id: &str, hash: u64) -> CacheState {
+        let cur_ts = utils::get_epoch_ms() - self.repub_interval;
+        let desync = self
+            .list
+            .values()
+            .any(|data| data.cache_hash != hash && data.publication_timestamp > cur_ts);
+
+        if desync {
+            CacheState::Desynced {
+                is_leader: !self.list.values().any(|data| {
+                    data.cache_hash == hash
+                        && data.peer_id.as_str() < peer_id
+                        && data.publication_timestamp > cur_ts
+                }),
+            }
+        } else {
+            CacheState::Synced
+        }
+    }
+}
+
+/// Join the dht and keep a local cache up to date
+///
+/// the resend interval is expressed in milliseconds
+pub fn cache_channel(
+    local: LocalCache,
+    swarm: Swarm<DomoBehaviour>,
+    resend_interval: u64,
+) -> (Cache, impl Stream<Item = DomoEvent>) {
+    let local_peer_id = swarm.local_peer_id().to_string();
+
+    let (cmd, r, _j) = dht_channel(swarm);
+
+    let peers_state = Arc::new(RwLock::new(PeersState::with_interval(
+        resend_interval as u128,
+    )));
+
+    let cache = Cache {
+        peers: peers_state.clone(),
+        local: local.clone(),
+        cmd: cmd.clone(),
+        peer_id: local_peer_id.clone(),
+    };
+
+    let stream = UnboundedReceiverStream::new(r);
+
+    let local_read = local.clone();
+    let cmd_update = cmd.clone();
+    let peer_id = local_peer_id.clone();
+
+    tokio::task::spawn(async move {
```

Comment:
I think that this could be transformed into a `spawn_local` without any issue, leading to the possibility of removing the `unsafe impl std::marker::Sync for InnerCache {}`

Reply:
We can try that way, I have a pending patch that moves the writes in a separate task using a channel, but if we can get away with a LocaSet it sounds overall cleaner.